File: 033/cpp/main.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int i = 0;
        int j = nums.size();
        while (i + 1 < j) {
            int k = i + (j - i) / 2;
            if (nums[i] < nums[k]) {
                if (target < nums[i]) {
                    i = k;
                } else if (target == nums[i]) {
                    return i;
                } else if (target < nums[k]) {
                    j = k;
                } else if (target == nums[k]) {
                    return k;
                } else {
                    i = k;
                }
            } else {
                if (target < nums[k]) {
                    j = k;
                } else if (target == nums[k]) {
                    return k;
                } else if (target < nums[i]) {
                    i = k;
                } else if (target == nums[i]) {
                    return i;
                } else {
                    j = k;
                }
            }
        }
        return nums[i] == target ? i : -1;
    }
};

File: 033/rust/main.rs
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let mut i = 0;
        let mut j = nums.len();
        while i + 1 < j {
            let k = i + (j - i) / 2;
            if nums[i] < nums[k] {
                if target < nums[i] {
                    i = k;
                } else if target == nums[i] {
                    return i as i32;
                } else if target < nums[k] {
                    j = k;
                } else if target == nums[k] {
                    return k as i32;
                } else {
                    i = k;
                }
            } else {
                if target < nums[k] {
                    j = k;
                } else if target == nums[k] {
                    return k as i32;
                } else if target < nums[i] {
                    i = k;
                } else if target == nums[i] {
                    return i as i32;
                } else {
                    j = k;
                }
            }
        }
        return if nums[i] == target {i as i32} else {-1};
    }
}
